### ISSUE TITLE
Implement NeuralUCB for Non-Linear Reward Modeling

### ISSUE THREAD

**Issue by evanvolgas** [ashita-ai/conduit#34]
## Overview
Implement NeuralUCB to handle non-linear reward patterns that linear bandits (LinUCB) cannot capture.

## Motivation
Current LinUCB assumes **linear reward structure**:
```
reward = w^T * context + noise
```

**Real-world LLM routing is non-linear**:
- Code queries → certain models excel (interaction effects)
- Long context + creative → different optimal models (feature interactions)
- Token count × complexity → non-linear cost-quality trade-offs

### Algorithm: NeuralUCB
**Paper**: "Neural Contextual Bandits with UCB-based Exploration" (Zhou et al., 2020)

**Key Features**:
1. Neural network for reward estimation (replaces linear model)
2. UCB-based exploration with neural uncertainty
3. Gradient-based online learning
4. Handles feature interactions and non-linearities

### Architecture
```python
class NeuralUCB(ContextualBandit):
    def __init__(self):
        self.reward_network = nn.Sequential(
            nn.Linear(context_dim, 128),
            nn.ReLU(),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, 1)  # reward prediction
        )
        self.uncertainty_estimator = ...  # Bootstrapped ensemble or dropout
```

### Phase 1: Core Algorithm (Week 1-2)
- [ ] Create `conduit/algorithms/neural_ucb.py`
- [ ] Neural network architecture (PyTorch)
- [ ] UCB exploration with neural uncertainty
- [ ] Online gradient updates (streaming learning)
- [ ] Model checkpointing and persistence

### Phase 2: Uncertainty Estimation (Week 3)
- [ ] Bootstrap ensemble for uncertainty (5 networks)
- [ ] Variance-based exploration bonus
- [ ] Calibration on historical data
- [ ] Uncertainty visualization for debugging

### Phase 3: Integration (Week 4)
- [ ] Router integration with model selection
- [ ] Warm-start from LinUCB (transfer learning)
- [ ] Hybrid mode: LinUCB → NeuralUCB at 5,000 queries
- [ ] Resource management (GPU optional, CPU fallback)

### Phase 4: Evaluation (Week 5-6)
- [ ] Synthetic non-linear benchmarks
- [ ] A/B test vs LinUCB on production traffic
- [ ] Sample efficiency analysis
- [ ] Computational overhead measurement

## Expected Benefits
- **Better Modeling**: Capture complex feature interactions
- **Quality Improvement**: 10-20% better routing for complex queries
- **Adaptive**: Learn non-linear patterns from data
- **Flexible**: Can model arbitrary reward functions

## Trade-offs
**Pros**:
- Captures non-linear patterns
- Handles feature interactions
- More expressive than linear models

**Cons**:
- Higher computational cost (10-50ms per query)
- Requires more samples for convergence (10,000+ queries)
- More hyperparameters to tune
- Needs GPU for best performance (optional)

## Configuration
```python
router = Router(
    algorithm="neural_ucb",
    neural_config={
        "hidden_dims": [128, 64],
        "learning_rate": 0.001,
        "batch_size": 32,
        "ensemble_size": 5,
        "exploration_weight": 2.0,
        "device": "cpu"  # or "cuda"
    }
)
```

## Resource Requirements
- **CPU Mode**: +20-50ms latency per query
- **GPU Mode**: +5-10ms latency per query
- **Memory**: ~50MB for model weights (ensemble of 5)
- **Storage**: Periodic model checkpoints (~10MB each)

## Warm-Start Strategy
Don't deploy NeuralUCB from scratch - warm-start from LinUCB:

1. Run LinUCB for 5,000 queries (establish baseline)
2. Train NeuralUCB on LinUCB's historical data (supervised pre-training)
3. Switch to NeuralUCB with online learning
4. Continue improving from warm-start

**Expected**: Match LinUCB performance in 1,000 queries, exceed by 2,000 queries

## Testing Strategy
1. **Synthetic Data**: Non-linear reward functions (polynomial, XOR patterns)
2. **Ablation Study**: Linear vs non-linear features
3. **Benchmark**: Compare to LinUCB on same data
4. **Production A/B Test**: 10% traffic for 2 weeks

## Success Criteria
- [ ] Outperforms LinUCB by 10%+ on non-linear benchmarks
- [ ] Matches LinUCB on linear benchmarks (no regression)
- [ ] <100ms latency on CPU (P95)
- [ ] Converges in <15,000 queries (acceptable for quality gain)

## Dependencies
```
torch>=2.0.0
numpy>=1.24.0
scipy>=1.10.0
```

## Database Changes
```sql
-- Store neural network checkpoints
CREATE TABLE neural_checkpoints (
    id SERIAL PRIMARY KEY,
    algorithm VARCHAR(50),
    checkpoint_data BYTEA,  -- serialized model weights
    query_count INTEGER,
    performance_metrics JSONB,
    timestamp TIMESTAMP DEFAULT NOW()
);
```

## Migration Path
1. **Phase 1**: Deploy as opt-in (`algorithm="neural_ucb"`)
2. **Phase 2**: A/B test vs LinUCB (10% traffic)
3. **Phase 3**: Hybrid default (LinUCB → NeuralUCB at 5K queries)
4. **Phase 4**: NeuralUCB default for production (if successful)

## References
- Zhou et al. (2020): "Neural Contextual Bandits with UCB-based Exploration"
- Riquelme et al. (2018): "Deep Bayesian Bandits Showdown"
- Zhang et al. (2021): "Neural Thompson Sampling"

## Priority
**Medium** - Significant quality improvement potential, but higher complexity and resource requirements than other algorithms

## Estimated Effort
6 weeks (1 engineer) + GPU infrastructure setup
